src/main/scala/com/fulcrumgenomics/util/zip/Zipper.scala

```diff
@@ -0,0 +1,840 @@
+
+  /** Returns true if `isStopped` has been set, or all data has been read in, consumed, and written out. */
+  private def isDone: Boolean = {
+    val ret = if (isStopped)                    true // we've been stopped
+    else {
+      if (!reader.isStopped)                    false // reader has not completed
+      else if (toConsumeQueue.nonEmpty)         false // more blocks to consume
+      else if (toWriteQueue.nonEmpty)           false // more blocks to write
+      else if (unusedQueue.size() != numBlocks) false // not all blocks written out (since they should be returned)
+      else                                      true  // all done!
+    }
+    ret
+  }
+
+  def run(): Unit = {
+    // Create the threads
+    val threads = (Seq(reader, writer) ++ consumers).map { runnable => new Thread(runnable) }
+    threads.foreach { thread => thread.setUncaughtExceptionHandler(threadExceptionHandler) }
+
+    // start your engines
+    threads.foreach { thread => thread.start() }
+
+    while (!isDone) {
+      // let others do stuff for a bit
+      Thread.sleep(Zipper.TimeoutInMilliseconds)
+
+      // if the reader reaches an EOF, close the toConsumeQueue only if it is empty
+      if (!toConsumeQueue.isClosed && reader.isStopped && toConsumeQueue.isEmpty) toConsumeQueue.close()
+
+      // if all the consumers are done, close the toWriteQueue
+      if (!toWriteQueue.isClosed && !consumers.exists(c => !c.isStopped)) toWriteQueue.close()
+
+      /*
+      val total = toConsumeQueue.size() + toWriteQueue.size() + unusedQueue.size() + reader.blocksBuffered
+      logger.debug(" (U=" + unusedQueue.size() + ", C=" + toConsumeQueue.size() + ", W=" + toWriteQueue.size() +
+        ", R=" + reader.blocksBuffered + ")"
+      )
+      */
+    }
+
+    // throw the exception if one was raised in a thread
+    threadException.map { thr => throw thr }
+
+    // stop all runnables
+    require(reader.isStopped) // should be done, so need to stop it
+    consumers.foreach { consumer => consumer.stopIt() }
+    writer.stopIt()
+
+    // close the queues
+    unusedQueue.close()
+    toConsumeQueue.close()
+    toWriteQueue.close()
+
+    // join
+    threads.foreach { thread => thread.join() } // TODO: timeout?
+
+    // forces the writer to perform any last steps (ex. write the last empty GZIP block).
+    writer.finish()
+
+    // stop itself
+    stopIt()
+  }
+
+  /** Closes the input and output channels, and any other streams opened using them. */
+  def close(): Unit = {
+    reader.close()
+    writer.close()
+  }
+}
+
+/** Companion methods for input and output blocks of data. */
+private[zip] object Blocks {
+  /** Returns the approximate size in bytes for Blocks */
+  val sizeInBytes: Long = {
+    SequentialBlocks.sizeInBytes(true) + SequentialBlocks.sizeInBytes(false)
+  }
+
+  /** The default block length in bytes of uncompressed data. */
+  val DefaultUncompressedFixedBlockLength = DEFAULT_UNCOMPRESSED_BLOCK_SIZE
```

Comment:
I spent quite a bit of time trying to locate `DEFAULT_UNCOMPRESSED_BLOCK_SIZE` before finding the import.  Can you please prefix these values coming in from a separate library (htsjdk) with their class name?